pyperformance/data-files/benchmarks/bm_sphinx/run_benchmark.py

```diff
@@ -0,0 +1,137 @@
+"""
+Build a subset of Python's documentation using Sphinx
+"""
+
+import gc
+import io
+import os
+from pathlib import Path
+import shutil
+import sys
+
+import pyperf
+from sphinx.cmd.build import main as sphinx_main
+
+
+# Sphinx performs a lot of filesystem I/O when it operates. This can cause the
+# results to be highly variable. Instead, we pre-load all of the source files
+# and then monkeypatch "open" so that Sphinx is reading from in-memory
+# `io.BytesIO` and `io.StringIO` objects.
+
+
+DOC_ROOT = (Path(__file__).parent / "data" / "Doc").resolve()
+
+
+_orig_open = open
+
+
+preloaded_files = {}
+real_io = set()
+
+
+def read_all_files():
+    for filename in DOC_ROOT.glob("**/*"):
+        if filename.is_file():
+            preloaded_files[str(filename)] = filename.read_bytes()
+
+
+def open(
+    file,
+    mode="r",
+    buffering=-1,
+    encoding=None,
+    errors=None,
+    newline=None,
+    closefd=True,
+    opener=None,
+):
+    if isinstance(file, Path):
+        file = str(file)
+    if isinstance(file, str):
+        if "r" in mode and file in preloaded_files:
+            if "b" in mode:
+                return io.BytesIO(preloaded_files[file])
+            else:
+                return io.StringIO(preloaded_files[file].decode(encoding or "utf-8"))
+        elif "w" in mode:
+            if "b" in mode:
+                newfile = io.BytesIO()
+            else:
+                newfile = io.StringIO()
+            preloaded_files[file] = newfile
+            return newfile
+
+    if isinstance(file, str):
+        if DOC_ROOT in Path(file).parents:
+            real_io.add(file)
+
+    return _orig_open(
+        file,
+        mode=mode,
+        buffering=buffering,
+        encoding=encoding,
+        errors=errors,
+        newline=newline,
+        closefd=closefd,
+        opener=opener,
+    )
+
+
+__builtins__.open = open
+
+
+def replace(src, dst):
+    pass
+
+
+os.replace = replace
+
+
+def build_html(doc_root):
+    sys.argv = [
+        "sphinx-build",
+        "-b",
+        "html",
+        "-d",
+        str(doc_root / "build" / "doctrees"),
+        "-j",
+        "1",
+        "-Q",
+        "--fresh-env",
+        "--write-all",
+        str(doc_root),
+        str(doc_root / "build" / "html"),
+    ]
+    t0 = pyperf.perf_counter()
+    sphinx_main()
```

Comment:
Using the `dummy` builder avoids writing HTML, and the slowdown / creation of reference cycles is in the parsing (reading) step rather than writing. 

Also this passes arguments directly rather than patching `sys.argv`

```suggestion
    t0 = pyperf.perf_counter()
    sphinx_main([
        "--builder",
        "dummy",
        "--doctree-dir",
        str(doc_root / "build" / "doctrees"),
        "--jobs",
        "1",
        "--silent",
        "--fresh-env",
        "--write-all",
        str(doc_root),
        str(doc_root / "build" / "html"),
    ])
```